lib/3scale/backend/worker_async.rb

```diff
@@ -73,6 +73,10 @@ def process_all
           break if @shutdown
 
           semaphore.async { perform(job) }
+
+          # Wait for the last `@max_concurrent_jobs` tasks to finish and clean the barrier.
+          # Otherwise it produces a memory leak. More info: https://github.com/3scale/apisonator/pull/376
+          barrier.wait if barrier.size > semaphore.limit
```

Comment:
I assume we can use `semaphore.limit * 2` or just some other bigger number if we are concerned about reuced throughput. e.g. we can consider `1000` entries a small enough number to keep in memory while not disturbing the queue parallelization as much as waiting for each new batch. The value doesn't seem to really depend on the semaphore size as long as it is a bigger number.

If we want to squeeze out every last bit of performanec, we could also keep a counter in the loop and wait every `X` iterations of the loop. But that's over the top I think..

I mean keeping as is is fine for me. Just saying that there are other options.

Reply:
Yes, the batch size doesn't need to be related to the semaphore size at all. I took the semaphore size to reduce complexity and because tasks will start waiting when the semaphore is full anyway. I think I'm going to make some performance tests with different batch sizes and check if I notice any difference. If there are no differences in performance, I'll keep the batch set to `@max_concurrent_jobs` for simplicity.

Reply:
I made some tests and I haven't observed big differences in performance between different batch sizes. However, I found that sometimes the worker just stops processing jobs even when there are new jobs being added by the listener. I'll have to investigate this.

Reply:
hmm, could it be some kind of a dead lock, but why? Do you see it regardless of batch size?

Reply:
It looks like a dead lock. I saw it with different batch sizes

Reply:
What `hey` parameters did you use? I can't reproduce locally with `hey -z 30m -c 60`.